Use `mobj`s to pass shared buffers

lorc opened the pull request:
This patch series includes three patches:

 * added `msg_parm.c` to ease up REE-TEE parameters handling
 * added new `mobj` type `mobj_shm`. This `mobj` describes buffer in preshared memory region (` CORE_MEM_NSEC_SHM`).
 *  All shared buffers are wrapped into `mobjs` (both command buffers, and buffers allocated in `thread.c`). All users of shared buffers (e.g. user TA loader) now work with `mobj`s. This allows to pass another `mobj` types, like registered `mobj` which will be introduced in #1232

jenswi-linaro commented on core/include/optee_msg.h:
```diff
@@ -56,6 +56,19 @@
+ * registered memory buffers, but for shared memory buffers requested by
+ * OPTEE via RPC.
+ */
+#define OPTEE_MSG_ATTR_TYPE_NEXT_FRAGMENT	0xc
```
Obsolete define.

jenswi-linaro commented on core/tee/tee_fs_rpc.c:
```diff
@@ -335,10 +333,12 @@ TEE_Result tee_fs_rpc_rename(uint32_t id, struct tee_pobj *old,
+				     TEE_FS_NAME_MAX, cookie,
+				     MSG_PARAM_MEM_DIR_IN))
+		return TEE_ERROR_BAD_STATE;
+
```
Empty line not needed

jenswi-linaro commented on core/tee/tee_fs_rpc_cache.c:
```diff
@@ -28,19 +28,21 @@
+		thread_rpc_free_payload(tsd->rpc_fs_payload_cookie,
+					tsd->rpc_fs_payload_mobj);
 		tsd->rpc_fs_payload = NULL;
 		tsd->rpc_fs_payload_cookie = 0;
```
`tsd->rpc_fs_payload_mobj` should be cleared too.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,221 @@
+ *   page count on success
+ *   <0 on error
+ */
+ssize_t msg_param_extract_pages(struct optee_msg_param *params, paddr_t *pages,
```
This function isn't used yet, and in later versions it looks differently. Please remove.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,221 @@
+ *  struct shmem_mapping of mapped buffer on success
+ *  NULL on error.
+ */
+struct mobj *msg_param_map_buffer(paddr_t pa_params, size_t num_params)
```
This function isn't used. Please remove.

jenswi-linaro commented on core/kernel/msg_param.c:
```diff
@@ -0,0 +1,221 @@
+			     uint64_t cookie, enum msg_param_mem_dir dir)
+{
+	if (mobj_matches(mobj, CORE_MEM_REG_SHM)) {
+		/* Registered SHM mobj */
```
This isn't used yet, but I see no harm in introducing it now.

lorc commented:
@etienne-lms , thank you for review. Looks like most of your comments was for code that is removed now. But I'll address them in original PR #1232 

I have addressed all comments and also fixed travis build issue (`mobj_shm` was not treated as non-secure).

lorc commented:
@jenswi-linaro,  thanks. Should be fixed now.

jenswi-linaro commented on core/arch/arm/tee/pta_socket.c:
```diff
@@ -239,19 +242,21 @@ static TEE_Result socket_ioctl(uint32_t instance_id, uint32_t param_types,
+	if (msg_param_get_buf_size(msg_params + 1) <= params[1].memref.size)
+		memcpy(params[1].memref.buffer, va,
+		       msg_param_get_buf_size(msg_params + 1));
+	params[1].memref.size = msg_param_get_buf_size(msg_params + 1);
```
Any particular reason why this function doesn't use the same pattern to update `params[1].memref` as `socket_recv()` above?

lorc replied:
I just followed the same logic, that was in original code. Or I didn't get what are you asking about :)

Only optimization that I can image there - is to add temporary variable to store size, so `msg_param_get_buf_size()` will not be called three times in a row. That can improve readability, but actually will not optimize code.
Because, looking to `objdump` output I can see that compiler successfully inlined that call and also it determined that `param->u.tmem.size` and `param->u.rmem.size` points to the same memory, so it basically eliminated all extra logic (and extra calls to `msg_param_get_buf_size` also).

jenswi-linaro replied:
Ah, yes I see the inconsistency in the original code too. I suppose that answers my question.

jenswi-linaro commented on core/arch/arm/tee/entry_std.c:
```diff
@@ -341,11 +342,27 @@ static void entry_cancel(struct thread_smc_args *smc_args,
+	struct optee_msg_arg *arg;
+	size_t args_size;
+
+	arg = (struct optee_msg_arg *)phys_to_virt(parg, MEM_AREA_NSEC_SHM);
```
No need to cast, `phys_to_virt()` returns `void *`.

etienne-lms commented on core/arch/arm/kernel/thread.c:
```diff
@@ -593,6 +594,7 @@ void __thread_std_smc_entry(struct thread_smc_args *args)
 		tee_fs_rpc_cache_clear(&thr->tsd);
 		if (!thread_prealloc_rpc_cache) {
 			thread_rpc_free_arg(thr->rpc_carg);
+			mobj_free(thr->rpc_mobj);
```
suggested `thr->rpc_mobj = NULL`, but i could not insure this is really required.

jenswi-linaro replied:
Anyhow, it's good practice to not leave stale pointers around.

lorc commented:
Addressed last comments in v3 patches.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

lorc commented:
@jenswi-linaro, thank you.
I have squashed deltas, rebased on `master` and added R-B tag.

etienne-lms commented:
2 remaining comments in 1st commit about obsolete `OPTEE_MSG_ATTR_TYPE_NEXT_FRAGMENT`.
Clean now or later? Whatever, `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`.

lorc commented:
@etienne-lms, thanks. I have removed that chunk. Should be fine now.

jenswi-linaro commented:
`Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (Hikey)`

lorc commented:
@jenswi-linaro, thank you. Added your `T-B` tag.
